File: publite2/civlauncher.py
import subprocess
import time
import pwd
import os
import threading
import re
from threading import Thread
from pathlib import Path
from datetime import datetime
import logging


logger = logging.getLogger(__name__)


class Civlauncher(Thread):
    def __init__(self, gametype, scripttype, new_port, metahostpath, savesdir, shutdown_event=None):
        super().__init__()
        self.daemon = True  # Thread will not prevent program exit
        self.new_port = new_port
        self.gametype = gametype
        self.scripttype = scripttype
        self.metahostpath = metahostpath
        self.savesdir = Path(savesdir)
        self.started_time = datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S")
        self.num_start = 0
        self.num_error = 0
        self.shutdown_event = shutdown_event or threading.Event()

    def run(self):
        while not self.shutdown_event.is_set():
            try:
                logger.info(
                    "Start freeciv-web on port %s and freeciv-proxy on port %s.",
                    self.new_port,
                    1000 + self.new_port,
                )
                self.launch_game()
                self.num_start += 1
            except (OSError, subprocess.SubprocessError, ValueError) as e:
                logger.error("Error during execution: %s", e)
                self.num_error += 1
            except KeyboardInterrupt:
                logger.info("Received keyboard interrupt, stopping server on port %s.", self.new_port)
                break
            
            # Check shutdown event before sleeping
            if self.shutdown_event.wait(5):
                break
        
        logger.info("Civlauncher thread for port %s shutting down.", self.new_port)

    def launch_game(self):
        # Prepare save directory
        self.savesdir.mkdir(parents=True, exist_ok=True)

        # Build arguments
        args = self.build_freeciv_args()

        # Use absolute path for logs directory
        script_dir = Path(__file__).parent.resolve()
        logs_dir = script_dir.parent / "logs"
        logs_dir.mkdir(parents=True, exist_ok=True)

        # Set up environment with library path
        home_dir = Path(pwd.getpwuid(os.getuid()).pw_dir)
        env = os.environ.copy()
        lib_path = str(home_dir / "freeciv" / "lib")
        existing = env.get('LD_LIBRARY_PATH', '')
        env['LD_LIBRARY_PATH'] = f"{lib_path}:{existing}" if existing else lib_path

        proxy_process = None
        proxy_log_file = None
        freeciv_log_file = None
        
        try:
            # Start proxy process - keep file handle open for process lifetime
            proxy_log = logs_dir / f"freeciv-proxy-{1000 + self.new_port}.log"
            proxy_log_file = open(proxy_log, "w")
            proxy_process = subprocess.Popen(
                ["websockify", str(1000 + self.new_port), f"localhost:{self.new_port}"],
                stdout=proxy_log_file,
                stderr=subprocess.STDOUT,
            )
            logger.info("Proxy started on port %s.", 1000 + self.new_port)

            # Start Freeciv-web process - keep file handle open for process lifetime
            freeciv_log = logs_dir / f"freeciv-web-stderr-{self.new_port}.log"
            freeciv_log_file = open(freeciv_log, "w")
            freeciv_process = subprocess.Popen(
                args,
                env=env,
                stdout=subprocess.DEVNULL,
                stderr=freeciv_log_file,
            )
            freeciv_process.wait()
            logger.info("Freeciv-web process exited with code %s.", freeciv_process.returncode)

        finally:
            # Clean up proxy process
            if proxy_process:
                proxy_process.terminate()
                proxy_process.wait()
                logger.info("Proxy process terminated.")
            
            # Close file handles after processes are done
            if proxy_log_file:
                proxy_log_file.close()
            if freeciv_log_file:
                freeciv_log_file.close()

    def build_freeciv_args(self):
        # Get home directory from system user database (not from HOME env var)
        # This prevents command injection via HOME environment variable manipulation
        home_dir = Path(pwd.getpwuid(os.getuid()).pw_dir)
        freeciv_binary = home_dir / "freeciv" / "bin" / "freeciv-web"
        
        # Resolve to absolute path and validate it exists
        try:
            freeciv_binary = freeciv_binary.resolve(strict=True)
        except (OSError, FileNotFoundError) as e:
            logger.error(
                "Freeciv binary not found. Expected location: %s. Error: %s",
                freeciv_binary,
                e,
            )
            raise
        
        # Validate metahostpath to prevent command injection
        # Allow only alphanumeric, dots, colons, slashes, and dashes
        if not re.match(r'^[a-zA-Z0-9.:/_-]+$', self.metahostpath):
            logger.error(
                "Invalid metahostpath format: %s. Must contain only alphanumeric, dots, colons, slashes, and dashes.",
                self.metahostpath,
            )
            raise ValueError(f"Invalid metahostpath: {self.metahostpath}")
        
        # Use absolute path for logs directory
        script_dir = Path(__file__).parent.resolve()
        logs_dir = script_dir.parent / "logs"
        log_file = logs_dir / f"freeciv-web-log-{self.new_port}.log"
        
        args = [
            str(freeciv_binary),
            "--debug", "1",
            "--port", str(self.new_port),
            "--Announce", "none",
            "--exit-on-end",
            "--meta", "--keep",
            "--Metaserver", f"http://{self.metahostpath}",
            "--type", self.gametype,
            "--read", f"pubscript_{self.scripttype}.serv",
            "--log", str(log_file),
            "--quitidle", "20",
            "--saves", str(self.savesdir),
        ]
        return args

File: publite-go/civlauncher.go
package main

import (
	"fmt"
	"log/slog"
	"os"
	"os/exec"
	"os/user"
	"path/filepath"
	"regexp"
	"time"
)

var metaHostPathRe = regexp.MustCompile(`^[a-zA-Z0-9.:/_-]+$`)

// CivLauncher starts and restarts a single freeciv-web + websockify pair.
type CivLauncher struct {
	GameType     string
	ScriptType   string
	Port         int
	MetaHostPath string
	SavesDir     string
	shutdown     chan struct{}
	StartedTime  string
	NumStart     int
	NumError     int
}

// NewCivLauncher creates a CivLauncher ready to run.
func NewCivLauncher(gameType, scriptType string, port int, metaHostPath, savesDir string, shutdown chan struct{}) *CivLauncher {
	return &CivLauncher{
		GameType:     gameType,
		ScriptType:   scriptType,
		Port:         port,
		MetaHostPath: metaHostPath,
		SavesDir:     savesDir,
		shutdown:     shutdown,
		StartedTime:  time.Now().UTC().Format("2006-01-02 15:04:05"),
	}
}

// Run is the main goroutine loop; it restarts the game server until shutdown.
func (cl *CivLauncher) Run() {
	for {
		select {
		case <-cl.shutdown:
			slog.Info("CivLauncher shutting down", "port", cl.Port)
			return
		default:
		}

		slog.Info("Starting freeciv-web", "port", cl.Port, "proxy_port", 1000+cl.Port)
		if err := cl.launchGame(); err != nil {
			slog.Error("Error during execution", "port", cl.Port, "error", err)
			cl.NumError++
		} else {
			cl.NumStart++
		}

		select {
		case <-cl.shutdown:
			slog.Info("CivLauncher shutting down", "port", cl.Port)
			return
		case <-time.After(5 * time.Second):
		}
	}
}

func (cl *CivLauncher) launchGame() error {
	if err := os.MkdirAll(cl.SavesDir, 0o755); err != nil {
		return fmt.Errorf("creating saves dir: %w", err)
	}

	args, logsDir, err := cl.buildFreecivArgs()
	if err != nil {
		return err
	}

	if err := os.MkdirAll(logsDir, 0o755); err != nil {
		return fmt.Errorf("creating logs dir: %w", err)
	}

	// Start websockify proxy.
	proxyLogPath := filepath.Join(logsDir, fmt.Sprintf("freeciv-proxy-%d.log", 1000+cl.Port))
	proxyLogFile, err := os.Create(proxyLogPath)
	if err != nil {
		return fmt.Errorf("creating proxy log: %w", err)
	}
	defer proxyLogFile.Close()

	proxyCmd := exec.Command("websockify",
		"--no-ssl",
		"--cert=",
		fmt.Sprintf("%d", 1000+cl.Port),
		fmt.Sprintf("localhost:%d", cl.Port),
	)
	proxyCmd.Stdout = proxyLogFile
	proxyCmd.Stderr = proxyLogFile
	if err := proxyCmd.Start(); err != nil {
		return fmt.Errorf("starting websockify: %w", err)
	}
	defer func() {
		_ = proxyCmd.Process.Kill()
		_ = proxyCmd.Wait()
		slog.Info("Proxy process terminated", "proxy_port", 1000+cl.Port)
	}()
	slog.Info("Proxy started", "proxy_port", 1000+cl.Port)

	// Start freeciv-web.
	freecivLogPath := filepath.Join(logsDir, fmt.Sprintf("freeciv-web-stderr-%d.log", cl.Port))
	freecivLogFile, err := os.Create(freecivLogPath)
	if err != nil {
		return fmt.Errorf("creating freeciv log: %w", err)
	}
	defer freecivLogFile.Close()

	freecivCmd := exec.Command(args[0], args[1:]...)
	freecivCmd.Stdout = nil // discard
	freecivCmd.Stderr = freecivLogFile
	if err := freecivCmd.Start(); err != nil {
		return fmt.Errorf("starting freeciv-web: %w", err)
	}
	if err := freecivCmd.Wait(); err != nil {
		slog.Info("Freeciv-web process exited", "port", cl.Port, "error", err)
	} else {
		slog.Info("Freeciv-web process exited successfully", "port", cl.Port)
	}
	return nil
}

func (cl *CivLauncher) buildFreecivArgs() (args []string, logsDir string, err error) {
	u, err := user.Current()
	if err != nil {
		return nil, "", fmt.Errorf("getting current user: %w", err)
	}
	freecivBinary := filepath.Join(u.HomeDir, "freeciv", "bin", "freeciv-web")
	if _, err := os.Stat(freecivBinary); err != nil {
		return nil, "", fmt.Errorf("freeciv binary not found at %s: %w", freecivBinary, err)
	}

	// Validate metahostpath to prevent command injection.
	if !metaHostPathRe.MatchString(cl.MetaHostPath) {
		return nil, "", fmt.Errorf("invalid metahostpath %q: must be alphanumeric with dots, colons, slashes, or dashes", cl.MetaHostPath)
	}

	// Determine the executable's directory so we can find sibling files
	// (pubscript_*.serv) and place logs one level up.
	exePath, err := os.Executable()
	if err != nil {
		return nil, "", fmt.Errorf("getting executable path: %w", err)
	}
	exeDir := filepath.Dir(exePath)
	logsDir = filepath.Join(exeDir, "..", "logs")
	logFile := filepath.Join(logsDir, fmt.Sprintf("freeciv-web-log-%d.log", cl.Port))
	scriptFile := filepath.Join(exeDir, fmt.Sprintf("pubscript_%s.serv", cl.ScriptType))

	args = []string{
		freecivBinary,
		"--debug", "1",
		"--port", fmt.Sprintf("%d", cl.Port),
		"--Announce", "none",
		"--exit-on-end",
		"--meta", "--keep",
		"--Metaserver", fmt.Sprintf("http://%s", cl.MetaHostPath),
		"--type", cl.GameType,
		"--read", scriptFile,
		"--log", logFile,
		"--quitidle", "20",
		"--saves", cl.SavesDir,
	}
	// Logic Change: Only add --Ranklog for multiplayer games
	if cl.GameType == "multiplayer" {
		ranklogPath := fmt.Sprintf("../ranklog/ranklog-game-%d.score", cl.Port)
		args = append(args, "--Ranklog", ranklogPath)
	}
	return args, logsDir, nil
}
